Support for reindexing APM indices

tylersmalley opened the pull request:
Starting in 7.0, APM will be aligning with the property names defined in the [Elastic Common Schema](https://github.com/elastic/ecs). When APM users upgrade to 7.x, they will need to migrate those indices created prior to 7.0 for the data to be present in the APM app.

Blocked by the finalizing of mappings and the reindex script here: https://github.com/elastic/apm-integration-testing/pull/277

Discussion:

When we re-index, we are appending `-reindex-v7` to the current index name. With doing this, we will almost always run into conflicts with the APM template used to create the index in the first place.

By default, APM creates an index template named `apm-%{[beat.version]}` with an index pattern of `apm-%{[beat.version]}-*`. The default index names are as follows:

- `apm-%{[beat.version]}-sourcemap`
- `apm-%{[beat.version]}-error-%{+yyyy.MM.dd}`
- `apm-%{[beat.version]}-transaction-%{+yyyy.MM.dd}`
- `apm-%{[beat.version]}-span-%{+yyyy.MM.dd}`
- `apm-%{[beat.version]}-metric-%{+yyyy.MM.dd}`
- `apm-%{[beat.version]}-onboarding-%{+yyyy.MM.dd}`

Unfortunately, there is no way to ignore index templates on index creation or update an index pattern to not match indices ending in `-reindexed-v7`

There have been a few options which have been discussed which require input from the APM team. @elastic/apm-server @bleskes 

- Produce a warning with a list of possibly effecting templates prior to indexing, letting the user know they might create conflicts and they should remove them if they are no longer used or APM data for that version is no longer being ingested.
- If the index creation fails, we can give the user additional information on what to do. This could include a link to documentation on identifying and removing the effecting templates. It's also possible that we could produce a list of index templates by iterating over `/_template?filter_path=*.index_patterns` to see if any patterns match.
- Similar to how we stop/start ML/Watcher, we could remove the index templates and replace them once we are done with the re-index. This seems way too risky considering it might be a template they are using elsewhere which sets something like replicas. Additionally, they could still be ingesting data from the old APM Beat and removing the template will cause issues.
- The APM server can prompt for removal during the 7.x setup script.

Please also review the language used in the following screenshots and let us know if it should be changed.

Documentation links to: https://www.elastic.co/guide/en/apm/server/master/breaking-changes.html
More about ECS links to: https://github.com/elastic/ecs

![apm-list](https://user-images.githubusercontent.com/40265/52181624-b66b8700-27a8-11e9-910c-db9b4b5759dc.PNG)

When clicking on "Reindex" for an APM index, currently presented is a single warning which the user must check acknowledge the data will be transformed.
![apm-flyout](https://user-images.githubusercontent.com/40265/52181625-b8354a80-27a8-11e9-9b9a-a237706cbebd.PNG)


### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-operations

tylersmalley commented on src/legacy/core_plugins/apm_oss/index.js:
```diff
@@ -30,12 +32,24 @@ export default function apmOss(kibana) {
         indexPattern: Joi.string().default('apm-*'),
 
         // ES Indices
+        sourcemapIndices: Joi.string().default('apm-*'),
```
This was missing

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10378/)

joshdover commented on x-pack/plugins/upgrade_assistant/public/components/tabs/checkup/deprecations/list.tsx:
```diff
@@ -91,7 +88,7 @@ export const DeprecationList: StatelessComponent<{
       index: dep.index!,
       details: dep.details,
-      reindex: dep.message === OLD_INDEX_MESSAGE,
+      reindex: dep.reindex,
```
I think this map is probably not necessary anymore, could probably just cast `deprecations` to `IndexDeprecationDetails[]`

joshdover commented on x-pack/plugins/upgrade_assistant/public/components/tabs/checkup/deprecations/reindex/flyout/warnings_step.tsx:
```diff
@@ -107,6 +107,35 @@ export class WarningsFlyoutStep extends React.Component<
+              <p className="upgWarningsStep__warningDescription">
+                Starting in version 7.0.0, APM data will be represented in the Elastic Common
+                Schema. In order for legacy data to be included, it is required that the data is
+                re-indexed to support this new format.
```
I think we should probably say something here like "historical APM data is not visible until it is reindexed"

joshdover commented on x-pack/plugins/upgrade_assistant/server/lib/reindexing/reindex_service.ts:
```diff
@@ -250,6 +255,25 @@ export const reindexServiceFactory = (
   const startReindexing = async (reindexOp: ReindexSavedObject) => {
     const { indexName } = reindexOp.attributes;
+
+    const reindexBody = {
```
Think you need to use this `reindexBody` in the actual callCluster below?

joshdover commented on x-pack/plugins/upgrade_assistant/server/lib/apm/index.test.ts:
```diff
@@ -0,0 +1,117 @@
+  });
+});
+
+describe.only('isLegacyApmIndex', () => {
```
oops :)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10549/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10582/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10586/)

simitt commented:
Regarding template conflicts during reindexing: since we know there will most probably be conflicts while reindexing when using default settings, I suggest to either show the user upfront that they will need to clean up the templates or prompt for removal when the migration script is started. If feasible I'd prefer the options where we detect potential problematic templates and directly show them to the user, over pointing to an external documentation asking the user to query for the templates themselves. 

Regarding linking to the documentation, there is also a general [APM release notes](https://www.elastic.co/guide/en/apm/get-started/current/apm-release-notes.html) section. This section should give a high level overview of breaking changes for APM in general, and then link to the more detailed server section. It should also contain explanation around switching to ECS. I'd rather link to these more general APM docs from Kibana. 

> APM index needs converted to ECS format

How about we put the detail information around field changes, explaining why we move to ECS and linking to it, into the APM docs and only put some more general information into the migration assistant, e.g. `APM indices need conversion to 7.x format`.  

@bmorelli25 I'd appreciate your input regarding documentation linking and wording.

bmorelli25 commented:
> Regarding linking to the documentation, there is also a general APM release notes section. This section should give a high level overview of breaking changes for APM in general, and then link to the more detailed server section. It should also contain explanation around switching to ECS. I'd rather link to these more general APM docs from Kibana.

Yup, let's link to [this](https://www.elastic.co/guide/en/apm/get-started/master/apm-release-notes.html) release notes section from Kibana. From there, I'll handle the linking to other sections of the APM documentation with more details.

> How about we put the detail information around field changes, explaining why we move to ECS and linking to it, into the APM docs and only put some more general information into the migration assistant, e.g. APM indices need conversion to 7.x format.

I mostly agree with this. We should have a sentence (for the people who are curious, but not _that_ curious) and the bulk of information should be in the APM documentation. I also thinks it makes more sense to remove the "[More about ECS](https://github.com/elastic/ecs)" link from this page. That's something I can link to from the APM documentation. I think directing users to the APM documentation initially is more beneficial.

_In other words. Instead of this:_ 
> This index will be converted to ECS format
> Starting in version 7.0.0, APM data will be represented in the Elastic Common Schema. Historical APM data will not visible until it's reindexed.
> Documentation
> More about ECS

_I'm thinking something more like this:_

**This index will be converted to the 7.x format**
Starting in version 7.0.0, APM data will be represented in the Elastic Common Schema. Historical APM data will not visible until it's reindexed. More information on this change is available in the [APM release notes](https://www.elastic.co/guide/en/apm/get-started/master/apm-release-notes.html).

graphaelli commented:
We're working through an issue with conflicting field mapping, summarized with this example:

```
PUT _template/ex-6.6.0
{
  "index_patterns": [
    "ex-6.6.0-*"
  ],
  "mappings": {
    "properties": {
      "docker": {
        "properties": {
          "container": {
            "properties": {
              "image": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          }
        }
      },
      "container": {
        "properties": {
          "image": {
            "properties": {
              "name": {
                "type": "alias",
                "path": "docker.container.image"
              }
            }
          }
        }
      }
    }
  }
}
PUT ex-6.6.0-test
{
  "mappings": {
    "properties": {
      "container": {
        "properties": {
          "image": {
            "properties": {
              "name": {
                "ignore_above": 1024,
                "type": "keyword"
              }
            }
          }
        }
      }
    }
  }
}
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10809/)

tylersmalley commented:
Updated messaging and changed to be a warning instead of critical since critical denotes something which will prevent the cluster from starting.

![apm-list](https://user-images.githubusercontent.com/40265/52241829-9b247880-2889-11e9-81c6-c100f48ac6eb.PNG)


![apm-flyout](https://user-images.githubusercontent.com/40265/52241825-9364d400-2889-11e9-9734-01299b0dd1bc.PNG)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10842/)

tylersmalley commented:
@graphaelli should we move forward with merging this PR and treat the ES error as a bug?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11226/)

tylersmalley commented:
We decided to get around the index template conflicts by pre-pending to the index name, which did here: https://github.com/elastic/kibana/pull/30166
